appsec/appsec.go

```diff
@@ -34,20 +39,77 @@ func MonitorParsedHTTPBody(ctx context.Context, body interface{}) {
 	// bonus: use sync.Once to log a debug message once if AppSec is disabled
 }
 
+type userMonitoringError struct {
+	err    error
+	status struct {
+		grpc codes.Code
+		http int
+	}
+}
+
+func (err *userMonitoringError) ShouldBlock() bool {
+	return err.status.grpc == codes.Aborted || err.status.http == 403
+}
+func (err *userMonitoringError) GRPCStatus() *status.Status {
+	return status.New(err.status.grpc, err.Error())
+}
+func (err *userMonitoringError) GRPCCode() codes.Code {
+	return err.status.grpc
+}
+func (err *userMonitoringError) HTTPCode() int {
+	return err.status.http
+}
+func (err *userMonitoringError) Error() string {
+	return err.err.Error()
+}
+
+// SetUser wraps tracer.SetUser().
+// On top of associating user information to spans, it performs a WAF check
+// and returns an error in case `id` matches a blocked user ID in the AppSec
+// WAF security rules. It is recommended to use this SDK whenever dealing with
+// authenticated user requests and block that request's execution in case this function
+// returns an error. See the examples for more information.
+func SetUser(ctx context.Context, id string, opts ...tracer.UserMonitoringOption) error {
+	s, ok := tracer.SpanFromContext(ctx)
+	if !ok {
+		log.Debug("appsec: could not retrieve span from context. User ID tag won't be set")
+		return nil
+	}
+	tracer.SetUser(s, id, opts...)
+	if !appsec.Enabled() {
+		log.Debug("appsec: not enabled. User blocking checks won't be performed.")
+		return nil
+	}
+	if sharedsec.MonitorUser(ctx, id) {
+		if s, ok := tracer.SpanFromContext(ctx); ok {
+			s.SetTag("appsec.blocked", true)
+		}
+		return &userMonitoringError{
+			err: errors.New("Suspicious user detected. Associated requests should be blocked"),
+			status: struct {
+				grpc codes.Code
+				http int
+			}{grpc: codes.Aborted, http: 403},
```

Comment:
⚠️ Hard-coded values: I suggest you drop the status field for now.
It also reveals that your current "waf action abstraction" will have to be improved to communicate that information down here.